### ISSUE TITLE
Build System Revamp

### ISSUE THREAD

**Issue by arshaw** [fullcalendar/fullcalendar#3297]
To benefit both `fullcalendar` and `fullcalendar-scheduler`.

The most immediate goal is to get a unified set of testing utilities to be uses across both repos. I've found writing tests for fullcalendar to be very cumbersome, many times taking longer than writing the actual feature.

I've needed to do dirty things like this
https://github.com/fullcalendar/fullcalendar/blob/master/tests/automated/businessHours.js#L189-L278
which derives from
https://github.com/fullcalendar/fullcalendar-scheduler/blob/master/tests/util/geom.coffee
and
https://github.com/fullcalendar/fullcalendar-scheduler/blob/master/tests/util/time-grid.coffee

and these util directories are a mess:
https://github.com/fullcalendar/fullcalendar/tree/master/tests/lib
https://github.com/fullcalendar/fullcalendar-scheduler/tree/master/tests/util

**I can be the one to unify these test utils**, or more precisely, have the tools in scheduler extend from those in the core, but I don't have the initial setup to do so.

Working on the build system has always been very fun for me, but I also feel very guilty doing it, considering my time is often better spent writing features, so I'm hoping to get some community involvement on this one.

There are a number of fun build-system mini-projects that can be worked on order to achieve this goal and improve both projects in general.
## 1. Move from Bower to NPM (DONE)

Both projects will continue to be distributed as both Bower and NPM packages, but internally, I want to start using NPM only.

There are a number of appealing reasons to do this, but the main reason is that the previously mentioned test utilities could be accessed as CommonJS modules when coupled with Webpack...
## 2. Use Webpack with Karma

Once NPM is set up, `fullcalendar-scheduler` would depend on the `fullcalendar` project, which would include the `tests/lib/*.js` files into the NPM-published package, which could be accessed as CommonJS modules from scheduler's tests.

Would need a preprocessor for karma:

https://github.com/webpack/karma-webpack#alternative-usage
(I like the "Alternative Usage" better)

We would want to introduce Webpack into the Karma setup for both fullcalendar and fullcalendar-scheduler. With fullcalendar, we could eventually move the `src/` to ES6 and CommonJS modules (but not yet).
## 3. Move from Grunt to Gulp (DONE)

For the fullcalendar-scheduler project.

Also, break down the build tasks into individual files that live in a `tasks/` directory.

_Lower priority._
## 4. Update directory structure (partially DONE)

In both projects:
- convert `buid/` directory to `bin/` and `tasks/` (DONE)
- move archive-related temp files into `tmp/` (DONE)

For fullcalendar:
- move manual tests that live in `tests/*.html` to `tests/manual/` (DONE)
- move automated tests hat live in `tests/automated/` to `tests/`

_Do as final step_. To avoid merge conflicts throughout the refactor.

**Comment by arshaw** [comment]
hey @avindra you did a really solid job on the Travis integration. any interest in any part of this?

**Comment by avindra** [comment]
@arshaw: I'm interested in the Webpack build / ES6 conversion. It's becoming a very common pattern to write ES6 and publish as ES5 via a module bundler. I think authors are also starting to publish both the ES5 and ES6 code.

With an ES6 version published (within the same npm package, just in a different folder specified by `jsnext:main` or some other method),  developers consuming libraries can import only the code they need and have their bundler "tree-shake" the code they don't (producing a much smaller build). 

It's an interesting approach because it will let library maintainers continue adding as many features as they want, without having to have all their consumers compromise on storage / bandwidth costs for the features they don't need.

**Comment by arshaw** [comment]
@avindra, yes I would love to get the main project source code under ES6 and use proper modules. Would require some rearchitecting. Would also be interesting discovering how to decouple things in order to achieve the tree-shake. For example, lazy-loading the calendar Views might be hard because they are specified via string options in the main Calendar. Nothing we can't get around.

Getting the basic Webpack setup is the first step here. It will more immediately benefit the reusable testing utilities.

I plan on doing steps 1 and 3 this weekend. If you'd like to tackle step 2, regardless of my progress, I'd start building it as a Gulp task.

**Comment by arshaw** [comment]
I made a bunch or progress on these items: https://github.com/fullcalendar/fullcalendar/tree/build-system

I updated the task description to reflect what's done.

**Comment by akivajgordon** [comment]
First of all, fantastic library – thanks!

I was interested in contributing, but I had some difficulties getting set up, and I feel that with some work, we could improve the dev experience to encourage more contributions from others.

While reading the contributing guide, I noticed that the first step is to `npm install -g gulp-cli`. It seems to me like an unnecessary burden to place on the contributor, so that is where I'd like to start.

npm supports [run scripts](https://docs.npmjs.com/cli/run-script) through its CLI. If there was a run script for each of the public Gulp tasks, I think it would have the benefits of:

a) avoiding the need for newcomers to install Gulp
b) allowing the package to control the Gulp CLI version being used

Basically, the idea is putting Gulp behind the npm interface to hide that detail from the contributor. It has an additional benefit of being able to remove Gulp altogether (if an alternative tool is available, e.g. GNU Make) without affecting the developer's experience.

All that said, you may have already had a specific reason for using the gulp-cli as the task running interface, so if you did, I'd love to hear those reasons.

What do you think?

**Comment by arshaw** [comment]
hi @akivajgordon i recently started moving in that direction:
https://github.com/fullcalendar/fullcalendar/blob/3fba0824f026c21393eb89d5608cf8c81448a909/package.json#L93-L96

if you can think of any enhancements, please open a new ticket and we'll talk about it

**Comment by arshaw** [comment]
lots and lots of this has already been done. esp since moving to a monorepo (in master). closing. will write a blog post or something soon
